server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedInboundAdapter.java

```diff
@@ -0,0 +1,217 @@
+package org.infinispan.server.memcached;
+
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.server.memcached.logging.Log;
+import org.infinispan.server.memcached.logging.MemcachedAccessLogging;
+import org.infinispan.util.concurrent.CompletionStages;
+import org.infinispan.util.logging.LogFactory;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelProgressiveFuture;
+import io.netty.channel.ChannelProgressiveFutureListener;
+import io.netty.channel.ChannelProgressivePromise;
+import io.netty.util.AttributeKey;
+
+public class MemcachedInboundAdapter extends ChannelInboundHandlerAdapter {
+
+   static final AttributeKey<ByteBufPool> ALLOCATOR_KEY = AttributeKey.valueOf("allocator");
+   protected final static int MINIMUM_BUFFER_SIZE;
+
+   protected final static Log log = LogFactory.getLog(MemcachedInboundAdapter.class, Log.class);
+
+   private final MemcachedBaseDecoder decoder;
+
+   protected ByteBuf outbound;
+   protected ChannelProgressivePromise progressive;
+
+   private long progressiveStart;
+   private long progressiveEnd;
+
+   // Variable to resume auto read when channel can be written to again. Some commands may resume themselves after
+   // flush and may not want to also resume on writability changes
+   protected boolean resumeAutoReadOnWritability;
+
+   static {
+      MINIMUM_BUFFER_SIZE = Integer.parseInt(System.getProperty("infinispan.memcached.minimum-buffer-size", "4096"));
+   }
+
+   public MemcachedInboundAdapter(MemcachedBaseDecoder decoder) {
+      this.decoder = decoder;
+   }
+
+   public static ByteBufPool getAllocator(ChannelHandlerContext ctx) {
+      ByteBufPool allocator = ctx.channel().attr(ALLOCATOR_KEY).get();
+      if (allocator == null) throw new IllegalStateException("Context does not have a buffer allocator");
+      return allocator;
+   }
+
+   private void flushBufferIfNeeded(ChannelHandlerContext ctx, boolean runOnEventLoop, MemcachedResponse res) {
+      if (outbound != null) {
+         if (runOnEventLoop) {
+            ctx.channel().eventLoop().execute(() -> {
+               if (res != null && progressive != null) {
+                  res.flushed(ctx.writeAndFlush(outbound));
+               } else {
+                  ctx.writeAndFlush(outbound, ctx.voidPromise());
+                  notifyPendingResponses(ctx);
```

Comment:
Don't pending responses require progressive? It seems inconsistent that progressive is not used in the other if branch but it used here.

Reply:
Updated.